cli_tools/google-osconfig-agent/ospatch/patch_run.go

```diff
@@ -356,50 +345,75 @@ func ackPatch(ctx context.Context, patchJobName string) {
 	// the server may have inadvertantly notified us twice (at least once deliver) so we
 	// can ignore it.
 	if currentPatchJob == nil || currentPatchJob.Job.PatchJobName != patchJobName {
-		res, err := reportPatchDetails(ctx, patchJobName, osconfigpb.Instance_NOTIFIED, 0, "")
-		if err != nil {
+		j := &patchJob{&osconfigpb.ReportPatchJobInstanceDetailsResponse{PatchJobName: patchJobName}}
+		pr := &patchRun{ctx: ctx, Job: j}
+		if err := pr.createClient(); err != nil {
+			logger.Errorf("Error creating osconfig client: %v", err)
+		}
+		if err := pr.reportPatchDetails(osconfigpb.Instance_NOTIFIED, 0, ""); err != nil {
 			logger.Errorf("reportPatchDetails Error: %v", err)
+			pr.close()
 			return
 		}
-		tasker.Enqueue("Run patch", func() {
-			patchRunner(ctx, &patchRun{Job: &patchJob{res}})
-		})
+		pr.PatchStep = acked
+		tasker.Enqueue("Run patch", pr.runPatch)
 	}
 }
 
-func reportPatchDetails(ctx context.Context, patchJobName string, patchState osconfigpb.Instance_PatchState, attemptCount int64, failureReason string) (*osconfigpb.ReportPatchJobInstanceDetailsResponse, error) {
-	// TODO: add retries. Patching shouldn't continue if we can't talk to the server.
-	logger.Debugf("Reporting patch details name:'%s', state:'%s', failReason:'%s'", patchJobName, patchState, failureReason)
-
-	client, err := osconfig.NewClient(ctx, option.WithEndpoint(config.SvcEndpoint()), option.WithCredentialsFile(config.OAuthPath()))
-	if err != nil {
-		logger.Errorf("osconfig.NewClient Error: %v", err)
-		return nil, err
+// retry tries to retry f for no more than maxRetryTime.
+func retry(maxRetryTime time.Duration, desc string, f func() error) error {
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	var tot time.Duration
+	for i := 1; ; i++ {
+		err := f()
+		if err == nil {
+			return nil
+		}
+		logger.Debugf("Error %s, attempt %d: %v", desc, i, err)
+
+		// Always increasing with some jitter, longest wait will be 5min + i seconds.
+		nf := math.Min(float64(i)*float64(i)+float64(rnd.Intn(i)), 300)
+		ns := time.Duration(int(nf)) * time.Second
+		tot += ns
+		if tot < maxRetryTime {
+			return err
+		}
```

Comment:
consider logging before sleeping. If the client sleeps for a few minutes, the last line in the log will say how long the client is sleeping for which can be helpful for debugging.